src/register-service-worker.js

```diff
@@ -0,0 +1,162 @@
+/* eslint-env browser */
+
+(function() {
+  var constants = require('./lib/constants.js');
+
+  var swScript = document.currentScript.dataset.serviceWorker;
+  var manifestAttribute = document.documentElement.getAttribute('manifest');
+
+  if (swScript && manifestAttribute && 'serviceWorker' in navigator) {
+    var manifestUrl = (new URL(manifestAttribute, location.href)).href;
+
+    var idb = require('idb');
+
+    idb.open(constants.DB_NAME, constants.DB_VERSION, function(upgradeDB) {
+      if (upgradeDB.oldVersion === 0) {
+        Object.keys(constants.OBJECT_STORES).forEach(function(objectStore) {
+          upgradeDB.createObjectStore(constants.OBJECT_STORES[objectStore]);
+        });
+      }
+    }).then(function(db) {
+      return Promise.all([
+        updateManifestAssociationForCurrentPage(db, manifestUrl),
+        handlePossibleManifestUpdate(db, manifestUrl)
+      ]);
+    }).then(function() {
+      return navigator.serviceWorker.register(swScript);
+    });
+  }
+
+  /**
+   * Caches the Responses for one or more URLs, using the Cache Storage API.
+   *
+   * @param {Array.<String>} urls
+   * @returns {Promise.<T>}
+   */
+  function addToCache(urls) {
+    return caches.open(constants.CACHE_NAME).then(function(cache) {
```

Comment:
This is going to introduce compatibility issues, unsure if it's ok.

In Appcache, there's a cache per manifest per version. If two pages use a different manifest, they use different caches.

If you load two tabs for the same page at the same time (therefore have the same manifest) they'll use the same cache.

If an updated manifest is detected, and successfully installed, a new cache is created for it. Now, if you load one of the above tabs, it'll be using the new cache, while the other tab continues to use the old one, unless it calls `swapCache()`.

Not sure how far down the rabbit hole we want to go here. GC is definitely an issue, as you can only clear caches once they've been superseded & no tab is using the old version.

You could ignore this, and ditch the old cache as soon as some tab uses the new cache. This would do the right thing when updating appcache and refreshing one page, but would break with multiple tabs, as caches the other tabs were using would be gone.

Alternatively you could set up a `postMessage` thing where the SW asks its clients which cache they're using, and use that to prune caches that have been superseded & are no longer used.